os/drivers/usbhost/uvc/usbhost_uvc_ctrl.c

```diff
@@ -0,0 +1,1357 @@
+ *
+ * Input Parameters:
+ *   chain - reference to the video chain object.
+ *   rollback - flag to determine rollback required or not
+ *   xctrl - control information to be set on device.
+ *   xctrls_count - count of controls to be updated
+ *
+ * Returned Values:
+ *   OK on success, appropriate error code on failures
+ *
+ ****************************************************************************/
+int uvc_ctrl_commit(struct uvc_video_chain *chain, int rollback, const struct v4l2_ext_control *xctrls, unsigned int xctrls_count)
+{
+	unsigned int i;
+	struct uvc_control *ctrl;
+	struct uvc_entity *entity = queue_entry(chain->entities.head, struct uvc_entity, chain);
+	int ret = OK;
+
+	/* Find the control. */
+	for (; &entity->chain != &chain->entities; entity = queue_entry(entity->chain.head, struct uvc_entity, chain)) {
+		if (entity == NULL) {
+			continue;
+		}
+
+		for (i = 0; i < entity->ncontrols; ++i) {
+			ctrl = &entity->controls[i];
+			if (!CTRL_IS_SUPPORTED(ctrl->status)) {
+				continue;
+			}
+
+			if (ctrl->info.flags & F_AUTO || !(ctrl->info.flags & F_GET_CUR)) {
+				CTRL_CLEAR_LOADED(ctrl->status);
+			}
+
+			if (!CTRL_IS_DIRTY(ctrl->status)) {
+				continue;
+			}
+
+			if (!rollback) {
+				ret = uvc_query_ctrl(chain->dev, UVC_SET_CUR, ctrl->entity->id, chain->dev->intfnum, ctrl->info.selector, ctrl->cur, ctrl->info.size);
+			} else {
+				ret = OK;
+			}
+
+			if (rollback || ret < 0) {
+				memcpy(ctrl->cur, ctrl->bak, ctrl->info.size);
+			}
+
+			CTRL_CLEAR_DIRTY(ctrl->status);
+		}
+	}
+
+	usbhost_uvc_givesem(&chain->ctrl_mutex);
+	return ret;
+}
+
+/****************************************************************************
+ * Name: uvc_ctrl_get
+ *
+ * Description:
+ *   This function is used to get the control values from the camera device. If
+ *   already read from the device, this function will return the value from
+ *   control object instead to reading it again from the device.
+ *
+ *   This function is called from the uvc_get_ctrlvalue method of V4L2.
+ *
+ * Input Parameters:
+ *   chain - reference to the video chain object.
+ *   ctrl  - reference to the control object from which data has to be read.
+ *   ctrl_class - class the control command to be set.
+ *   xctrl - control information to be set on device.
+ *
+ * Returned Values:
+ *   OK on success, appropriate error code on failures
+ *
+ ****************************************************************************/
+int uvc_ctrl_get(struct uvc_video_chain *chain, struct uvc_control *ctrl, uint16_t ctrl_class, struct v4l2_ext_control *xctrl)
+{
+	int ret = OK;
+	int32_t val = 0;
+	unsigned int i;
+	struct uvc_control *control;
+	struct uvc_control_mapping *mapping;
+	const struct uvc_menu_info *menu = NULL;
+
+	if (ctrl == NULL) {
+		control = uvc_search_control(chain, xctrl->id, ctrl_class, &mapping);
+		if (control == NULL) {
+			uvdbg("Failed to find a valid control!!\n");
+			return -EINVAL;
```

Comment:
Is it right to return "-EINVAL" in the case where no control is found?
We could use ENODATA, however, that is generally used for data streams...

Reply:
I feel EINVAL is correct, as we trying to search something which is invalid for the current instance of UVC device.

Reply:
Ohkay